### ISSUE TITLE
Extended Support for PyTorch v0.4+

### ISSUE THREAD

**Issue by nasimrahaman** [inferno-pytorch/inferno#103]
This is an issue to track the next inferno release built around PyTorch 0.4. Below is a list of what is to come, feel free to populate it and/or suggest changes. 

## Core
- [x] Adapt to the Pytorch v0.4 paradigm and deprecate `Variables`,
- [x] Integrate zero-dimensional tensors (and get rid of all `variable.data[0]` in the codebase),
- [ ] Integrate the new device-agnostic constructs (`tensors.to(...)` or `model.to(...)`),
- [ ] Wrapper to manage gradient checkpointing,
- [ ] Integrate support for `reduce=False` in all inferno-managed losses functions, 

## Visualization
- [x] Replace the existing tensorboard backend with [lanpa/tensorboardX](https://github.com/lanpa/tensorboard-pytorch). 

## General
- [ ] Make the trainer class more modular without compromising on functionality. Break-up the oversized `Trainer` class to smaller classes to facilitate future support for multi-model trainers.

### 15 Aug 2018
To fully implement all 0.4+ features without bloating the codebase, we'd need to deprecate v0.3 and below, potentially invalidating a lot of code. I guess this can wait till v1.0.

**Comment by bstriner** [comment]
For checking 0.4 compatibility (although there are still many warnings):

https://travis-ci.org/inferno-pytorch/inferno/builds/380847555?utm_source=github_status&utm_medium=notification
https://github.com/inferno-pytorch/inferno/pull/109

**Comment by constantinpape** [comment]
This is all implemented.
